# Exercici 1

Crea almenys dos models de regressió diferents per intentar predir el millor possible el preu de les vivendes (MEDV) de l'arxiu adjunt.

En primer lloc carregarem el nostre dataframe.

In [2]:
import pandas as pd # Importem la llibreria Pandas

df = pd.read_csv('housing data.csv') # Carreguem el dataframe
df.head() # Mostrem els primers resultats


0.00632  18.00  2.310  0  0.5380  6.5750  65.20  4.0900  1  296.0  15.30  \
0  0.02731    0.0   7.07  0   0.469   6.421   78.9  4.9671  2  242.0   17.8   
1  0.02729    0.0   7.07  0   0.469   7.185   61.1  4.9671  2  242.0   17.8   
2  0.03237    0.0   2.18  0   0.458   6.998   45.8  6.0622  3  222.0   18.7   
3  0.06905    0.0   2.18  0   0.458   7.147   54.2  6.0622  3  222.0   18.7   
4  0.02985    0.0   2.18  0   0.458   6.430   58.7  6.0622  3  222.0   18.7   

   396.90  4.98  24.00  
0  396.90  9.14   21.6  
1  392.83  4.03   34.7  
2  394.63  2.94   33.4  
3  396.90  5.33   36.2  
4  394.12  5.21   28.7

Com podem veure, no tenim encapçalament. L'afegirem a partir de la informació que tenim al document 'housing data description'

In [3]:
# Creem una llista amb els noms de les columnes
noms_columnes = ["CRIM", "ZN", "INDUS", "CHAS", "NOX", "RM", "AGE", "DIS", "RAD", "TAX", "PTRATIO", "B", "LSTAT", "MEDV"]

# Afegim l'encapçalament amb el nom de les columnes al dataframe
df_housing = pd.read_csv('housing data.csv', header=None, names=noms_columnes)

# Mostrem el dataframe
df_housing.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

Afegim la informació proporcionada que descriu les variables:

   1. CRIM      taxa de delicte per càpita per municipi
   2. ZN        proporció de sòl residencial zonificat per a lots de més de 25.000 peus quadrats.
   3. INDUS     proporció d'hectàrees de negocis no minoristes per municipi
   4. CHAS      variable dummy Charles River (= 1 si la parcel·la toca el riu; 0 en cas contrari)
   5. NOX       concentració d'òxids nitrosos (parts per 10 milions)
   6. RM        nombre mitjà d'habitacions per habitatge 
   7. AGE       proporció d'unitats ocupades pel propietari construïdes abans de 1940
   8. DIS       distàncies ponderades a 5 oficines d'ocupació de Boston
   9. RAD       índex d'accessibilitat a carreteres radials
   10. TAX      taxa d'impost sobre béns immobles per cada $10.000
   11. PTRATIO  ràtio alumne-mestre per municipi
   12. B        1000(Bk - 0,63)^2 on Bk és la proporció de negres per municipi.
   13. LSTAT    % de la població de baix estatus.
   14. MEDV     Valor mitjà de les llars ocupades pels propietaris en milers de dòlars.

Per aquest tipus conjunt de dades, podem utilitzar diversos models de regressió per predir el valor mitjà de les cases (MEDV) en funció dels altres atributs. Per exemple, el model de Regressió Líneal Múltiple, els Arbres de Decisió i Random Forest, el Gradient Boosting, Regressió Ridge i Lasso, Regressió de Suport Vectorial, fins i tot les Xarxes Neuronals.

El model de Regressió Lineal Múltiple ens pot ajudar a predir el valor de la casa en funció de múltiples variables independents. 
Pel que fa als Arbres de Decisió i el Random Forest, ens poden ajudar a trobar relacions no lineals i interaccions entre les característiques. Amb el Gradient Boosting podem trobar millors resultats en comparació amb el Random Forest. Pel que fa als models de Regressió Ridge i Lasso, ens poden ajudar a prevenir el sobreajustament (overfitting) pel fet que apliquen tècniques de regularització si hi ha moltes característiques en el conjunt de dades. El model de Regressió de Suport Vectorial (SVR) ens pot ajudar a capturar relacions ni lineals en les nostres dades. Finalment, les Xarxes Neuronals es podrien aplicar per modelar relacions complexes entre les variables, encara que el nostre conjunt de dades no és gaire gran i és bastant específic.

Abans d'aplicar algun d'aquests models, ens caldrà fer un anàlisi exploratòria de les dades (distribució, outliers, correlacions) i fer un preprocessament (si cal estandarització, veure si hi ha valors faltants i crear característiques derivades d'altres). Un cop aplicat el model, caldrà fer una validació creuada (cross-validation) per avaluar l'eficàcia de cada model i evitar el sobreajustament. Posteriorment, farem una avaluació del model utilitzant mètriques com l'MSE (Error Quadràtic Mitjà), la desviació mitjana (RMSE) o l'error mitjà (MAE). Finalment, seleccionarem el model segons el seu rendiment durant la validació creuada i l'avaluació en el conjunt de dades de prova.

Entre els diferents models descrits, decidim aplicar el model de Regressió Lineal Múltiple, el Random Forest i el Gradient Boosting.

Anem a veure el tipus de dades i si n'hi ha de faltants:

In [4]:
df_housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    int64  
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB


Com podem veure, tenim un total de 506 registres i no hi ha dades faltants en cap de les 14 variables. Pel que fa al tipus de dades, totes les variables són numèriques. La variables 'CHAS' i 'RAD' són nombres enters (int64) i la resta són nombres decimals (float64).

In [5]:
# En primer lloc farem un anàlisi estadístic de les dades

df_housing.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT        MEDV  
count  506.000000  506.000000  
mean    12.653063   22.532806  
std      7.141062    9.197104  
min      1.730000    5.000000  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

Com que moltes de les variables no segueixen una distribució normal, detectarem els outliers a partir del Isolation Forest, un algoritme de detecció d'anomalies que no requereix una distribució normal de les dades.

In [6]:
from sklearn.ensemble import IsolationForest
from sklearn.preprocessing import StandardScaler
import pandas as pd

# Escalarem les dades abans d'aplicar Isolation Forest
scaler = StandardScaler()
scaled_data = scaler.fit_transform(df_housing)

# Instanciem Isolation Forest
iso = IsolationForest(n_estimators=100, contamination=0.01, random_state=42)

# Ajustem les dades escalades
iso.fit(scaled_data)

# Obtenim les prediccions de l'algoritme (1 per inliers, -1 per outliers)
labels = iso.predict(scaled_data)

# Afegim les etiquetes al DataFrame original per marcar els outliers
df_housing['outlier'] = labels

# Filtrem el DataFrame per mostrar només els outliers
outliers_df = df_housing[df_housing['outlier'] == -1]

print(outliers_df) # Mostrem els outliers

         CRIM    ZN  INDUS  CHAS    NOX     RM    AGE     DIS  RAD    TAX  \
155   3.53501   0.0  19.58     1  0.871  6.152   82.6  1.7455    5  403.0   
283   0.01501  90.0   1.21     1  0.401  7.923   24.8  5.8850    1  198.0   
369   5.66998   0.0  18.10     1  0.631  6.683   96.8  1.3567   24  666.0   
370   6.53876   0.0  18.10     1  0.631  7.016   97.5  1.2024   24  666.0   
372   8.26725   0.0  18.10     1  0.668  5.875   89.6  1.1296   24  666.0   
414  45.74610   0.0  18.10     0  0.693  4.519  100.0  1.6582   24  666.0   

     PTRATIO       B  LSTAT  MEDV  outlier  
155     14.7   88.01  15.02  15.6       -1  
283     13.6  395.52   3.16  50.0       -1  
369     20.2  375.33   3.73  50.0       -1  
370     20.2  392.05   2.96  50.0       -1  
372     20.2  347.88   8.88  50.0       -1  
414     20.2   88.27  36.98   7.0       -1  


Anàlisi dels resultats: <br>

- CRIM: La taxa de criminalitat per capita varia significativament en aquestes observacions, amb un outlier notablement alt (45.74610).
- ZN: Proporció de terreny residencial zonificat per lots de més de 25.000 peus quadrats. Per a la majoria dels outliers, aquest valor és 0.0, indicant que potser són zones d'alta densitat on no hi ha zones grans zonificades per residències.
- INDUS: Proporció d'hectàrees de negocis no minoristes per ciutat. Alguns outliers estan en zones altament industrialitzades (18.10).
- CHAS: Variable dummy de Charles River (1 si la tracta toca el riu; 0 en cas contrari). En aquestes observacions, alguns dels outliers són propietats que toquen el riu.
- RM: Nombre mitjà d'habitacions per habitatge. Variacions en aquesta característica també poden influir en la classificació com a outlier.
- AGE: Proporció d'unitats ocupades pels propietaris construïdes abans de 1940. Observem que alguns outliers tenen propietats molt antigues (82.6 a 100.0).
- TAX: Taxa d'impostos a la propietat de valor total per \$10,000. Alguns outliers tenen taxes d'impostos alts (666.0).
- MEDV: Valor mitjà de les cases ocupades pels propietaris en $1000's. Hi ha una variabilitat aquí, amb algunes cases amb valors molt alts (50.0, que pot ser un valor capat per a la màxima mediana de preu de l'habitatge en aquest dataset).

En vista d'aquests resultats,  com que tenim diverses variables amb un nombre elevat d'outliers i algunes de les variables no presenten una distribució normal i sí força esbiaixament, decidim aplicar un Robut Scaler, que no és tant sensible als outliers com el Standard Scaler.

In [7]:
from sklearn.preprocessing import RobustScaler

# Instanciem l'objecte RobustScaler
scaler = RobustScaler()

# Seleccionem totes les variables independents, sense la variable objectiu 'MEDV'
features = df_housing.drop('MEDV', axis=1)

# Ajustem i transformem les dades (excepte la variable objectiu)
scaled_features = scaler.fit_transform(features)

# Convertim les característiques escalades de nou en un DataFrame per facilitar la manipulació i visualització
df_housing_scaled = pd.DataFrame(scaled_features, index=features.index, columns=features.columns)

# Observemn el dataframe amb les dades escalades
print(df_housing_scaled.head())


       CRIM    ZN     INDUS  CHAS       NOX        RM       AGE       DIS  \
0 -0.069593  1.44 -0.571650   0.0  0.000000  0.496612 -0.250765  0.285777   
1 -0.063755  0.00 -0.202943   0.0 -0.394286  0.287940  0.028542  0.569789   
2 -0.063760  0.00 -0.202943   0.0 -0.394286  1.323171 -0.334353  0.569789   
3 -0.062347  0.00 -0.581720   0.0 -0.457143  1.069783 -0.646279  0.924391   
4 -0.052144  0.00 -0.581720   0.0 -0.457143  1.271680 -0.475025  0.924391   

    RAD       TAX   PTRATIO         B     LSTAT  outlier  
0 -0.20 -0.087855 -1.339286  0.261902 -0.637681      0.0  
1 -0.15 -0.227390 -0.446429  0.261902 -0.221889      0.0  
2 -0.15 -0.227390 -0.446429  0.066675 -0.732634      0.0  
3 -0.10 -0.279070 -0.125000  0.153016 -0.841579      0.0  
4 -0.10 -0.279070 -0.125000  0.261902 -0.602699      0.0  


A continuació crearem el Model de Regressió Lineal Múltiple i evaluarem els seus resultats amb MSE i R2.

In [8]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Separem les característiques independents (X) i la variable dependent (y)
X = df_housing.drop('MEDV', axis=1)  # Eliminem 'MEDV' per tenir només variables independents
y = df_housing['MEDV']

# Dividim les dades en conjunts d'entrenament i test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Creem l'objecte de regressió lineal
model = LinearRegression()

# Entrenem el model utilitzant les dades d'entrenament
model.fit(X_train, y_train)

# Fem prediccions utilitzant les dades de test
y_pred = model.predict(X_test)

# Avaluem el model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f'Error Quadràtic Mitjà (MSE): {mse}')
print(f'Coeficient de Determinació (R^2): {r2}')

# Podem veure els coeficients del model per entendre la importància de cada variable
print('Coeficients del model:', model.coef_)


Error Quadràtic Mitjà (MSE): 23.300514245551152
Coeficient de Determinació (R^2): 0.6822676638049429
Coeficients del model: [-1.10837081e-01  2.03017948e-02  4.03261870e-02  3.60019940e-01
 -1.73411601e+01  4.59732519e+00 -7.30359161e-03 -1.34974553e+00
  2.54554311e-01 -1.11080817e-02 -9.44019798e-01  1.34054937e-02
 -4.62134713e-01 -7.29419325e+00]


Ara aplicarem un model Random Forest:

In [9]:
# Importem les llibreries
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
from math import sqrt

# Separem les característiques independents (X) i la variable dependent (y)
X = df_housing.drop('MEDV', axis=1)  # Asumim que 'MEDV' no està escalada
y = df_housing['MEDV']

# Dividim les dades en conjunts d'entrenament i test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Creem l'objecte Random Forest Regressor
rf = RandomForestRegressor(n_estimators=100, random_state=42)

# Entrenem el model utilitzant les dades d'entrenament
rf.fit(X_train, y_train)

# Fem prediccions utilitzant les dades de test
y_pred = rf.predict(X_test)

# Calculem l'error quadràtic mitjà (MSE) i el coeficient de determinació (R2)
mse = mean_squared_error(y_test, y_pred)
r2_score = rf.score(X_test, y_test)

print(f"MSE: {mse}")
print(f"R2: {r2_score}")


MSE: 8.602580392156865
R2: 0.8826928051243473


Ara aplicarem un model Gradient Boosting:

In [10]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error, r2_score
from math import sqrt

# Separem les característiques independents (X) i la variable dependent (y)
X = df_housing.drop('MEDV', axis=1)  # Asumim que 'MEDV' no està escalada
y = df_housing['MEDV']

# Dividim les dades en conjunts d'entrenament i test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Creem l'objecte Gradient Boosting Regressor
# Pots ajustar els paràmetres com n_estimators, learning_rate, i max_depth, entre d'altres
gb = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42)

# Entrenem el model utilitzant les dades d'entrenament
gb.fit(X_train, y_train)

# Fem prediccions utilitzant les dades de test
y_pred = gb.predict(X_test)

# Calculem l'error quadràtic mitjà (MSE) i el coeficient de determinació (R2)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"MSE: {mse}")
print(f"R2: {r2}")


MSE: 6.863643755263377
R2: 0.9064054319922645


# Exercici 2

Compara’ls en base al MSE i al R2.

Tal i com hem vist anteriorment, hem calculat els valors del coeficient de determinació (R2) i l'error quadràtic mitjà (MSE) pe tal de comparar els resultats dels tres models que hem aplicat:  Regressió lineal múltiple, Random Forest i Gradient Boosting. 

El coeficient R2 mesura la proporció de la variància de la variable dependent que és previsible a partir de les variables independents. Un valor de R2 més alt indica un millor ajust del model. El MSE mesura la qualitat d'un estimador, és a dir, quan més petit és el MSE, millors són les prediccions del model.

- Regressió lineal múltiple: R2 = 0.6823 i MSE = 23.30

- Random Forest: R2 = 0.8827 i MSE = 8.6026

- Gradient Boosting: R2 = 0.9064 i MSE = 6.8636

Basats en aquesta resultats, podem veure que el model de Gradient Boosting supera tant el Random Forest com la regressió lineal múltiple. El Gradient Boosting té el coeficient R2 més alt, indicant que fa una millor feina capturant la variància dels preus de les vivendes (MEDV). A més, té el MSE més baix, el que suggereix que les seves prediccions estan més a prop dels valors reals.

El Random Forest també ofereix bons resultats, superant significativament la regressió lineal múltiple, la qual té el rendiment més baix dels tres models segons aquestes mètriques. Això podria ser degut a la naturalesa no lineal i complexa de les dades, que la regressió lineal múltiple no pot capturar tant eficaçment com els models basats en arbres.

# Exercici 3

Entrena’ls utilitzant els diferents paràmetres que admeten per intentar millorar-ne la predicció.

Per a la regressió lineal múltiple, no hi ha molts paràmetres per ajustar, ja que és un model relativament simple. No obstant això, es podriem aplicar transformacions logarítmiques o de potència per a les variables independents que no tenen una relació lineal dependent.

Pel que fa al Random Forest, hi ha diversos hiperparàmetres que podem ajustar:

- n_estimators: Nombre d'arbres en el bosc.
- max_features: El nombre màxim de característiques considerades per dividir un node.
- max_depth: Profunditat màxima de l'arbre.
- min_samples_split: El nombre mínim de mostres requerides per dividir un node intern.
- min_samples_leaf: El nombre mínim de mostres requerides en un node fulla.
- bootstrap: Mètode per a les mostres de bootstrap quan es construeixen els arbres.

In [11]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from scipy.stats import randint as sp_randint

# Instanciem el Model amb els Hiperparàmetres Inicials
rf = RandomForestRegressor(n_estimators=100, 
                           max_features='sqrt',  # Canviat de 'auto' a 'sqrt'
                           max_depth=None, 
                           min_samples_split=2, 
                           min_samples_leaf=1, 
                           bootstrap=True, 
                           random_state=42)

# Ajustem el model al conjunt d'entrenament
rf.fit(X_train, y_train)

# Avaluem el model
predictions = rf.predict(X_test)
mse = mean_squared_error(y_test, predictions)
print(f"MSE inicial: {mse}")

# Ajustem els hiperparàmetres

# Utilitzem el GridSearchCV
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_features': ['sqrt', 'log2'],  # Eliminat 'auto'
    'max_depth': [4, 6, 8, 10],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

CV_rfr = GridSearchCV(estimator=rf, param_grid=param_grid, cv=5)
CV_rfr.fit(X_train, y_train)
print(f"Millors paràmetres (GridSearchCV): {CV_rfr.best_params_}")

# I ara utilitzem el RandomizedSearchCV
param_dist = {
    'n_estimators': sp_randint(100, 500),
    'max_features': ['sqrt', 'log2'],  # Eliminat 'auto'
    'max_depth': sp_randint(3, 20),
    'min_samples_split': sp_randint(2, 11),
    'min_samples_leaf': sp_randint(1, 11)
}

random_search = RandomizedSearchCV(rf, param_distributions=param_dist, n_iter=100, cv=5, random_state=42)
random_search.fit(X_train, y_train)
print(f"Millors paràmetres (RandomizedSearchCV): {random_search.best_params_}")

# Avaluem el model ajustat
best_rf = random_search.best_estimator_
predictions_best = best_rf.predict(X_test)
mse_best = mean_squared_error(y_test, predictions_best)
print(f"MSE amb hiperparàmetres ajustats: {mse_best}")


MSE inicial: 12.377189980392139
Millors paràmetres (GridSearchCV): {'max_depth': 8, 'max_features': 'sqrt', 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 200}
Millors paràmetres (RandomizedSearchCV): {'max_depth': 9, 'max_features': 'log2', 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 476}
MSE amb hiperparàmetres ajustats: 11.487434801540934


Amb l'ajustament dels hiperparàmetres, el MSE ha millorat de 12.3772 a 11.4874, el que indica que els hiperparàmetres optimitzats han ajudat a millorar el rendiment del model de Random Forest.

GridSearchCV ha trobat que una profunditat màxima (max_depth) de 8, max_features com 'sqrt', un mínim de mostres per node fulla (min_samples_leaf) de 1, un mínim de mostres per dividir un node (min_samples_split) de 2, i 200 arbres (n_estimators) com la millor combinació per al model.

RandomizedSearchCV ha proposat una configuració lleugerament diferent amb una profunditat màxima de 9, max_features com 'log2', i un nombre significativament major d'arbres, 476.

El fet que el MSE amb els paràmetres ajustats mitjançant RandomizedSearchCV sigui més baix indica que els hiperparàmetres trobats per aquest mètode d'optimització són més eficaços per al nostre conjunt de dades.

Ara ajustarem els hiperparàmetres del Gradient Boosting:

In [12]:
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from scipy.stats import randint as sp_randint

# Instanciem el model amb els hiperparàmetres inicials
gbm = GradientBoostingRegressor(random_state=42)

# Ajustamem els hiperparàmetres amb GridSearchCV
param_grid_gbm = {
    'n_estimators': [50, 100, 200],
    'learning_rate': [0.01, 0.05, 0.1],
    'max_depth': [3, 4, 5],
    'min_samples_split': [2, 4, 6],
    'min_samples_leaf': [1, 2, 4],
    'subsample': [0.8, 0.9, 1.0]
}

CV_gbm = GridSearchCV(estimator=gbm, param_grid=param_grid_gbm, cv=5)
CV_gbm.fit(X_train, y_train)
print(f"Millors paràmetres (GridSearchCV): {CV_gbm.best_params_}")

# O utilitzant RandomizedSearchCV per a una cerca més ràpida
param_dist_gbm = {
    'n_estimators': sp_randint(30, 250),
    'learning_rate': [0.001, 0.01, 0.05, 0.1],
    'max_depth': sp_randint(2, 10),
    'min_samples_split': sp_randint(2, 11),
    'min_samples_leaf': sp_randint(1, 11),
    'subsample': [0.6, 0.7, 0.8, 0.9, 1.0]
}

random_search_gbm = RandomizedSearchCV(gbm, param_distributions=param_dist_gbm, n_iter=100, cv=5, random_state=42)
random_search_gbm.fit(X_train, y_train)
print(f"Millors paràmetres (RandomizedSearchCV): {random_search_gbm.best_params_}")

# Avaluem el model ajustat
best_gbm = random_search_gbm.best_estimator_
predictions_best_gbm = best_gbm.predict(X_test)
mse_best_gbm = mean_squared_error(y_test, predictions_best_gbm)
print(f"MSE amb hiperparàmetres ajustats: {mse_best_gbm}")


Millors paràmetres (GridSearchCV): {'learning_rate': 0.05, 'max_depth': 3, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 200, 'subsample': 0.8}
Millors paràmetres (RandomizedSearchCV): {'learning_rate': 0.1, 'max_depth': 2, 'min_samples_leaf': 7, 'min_samples_split': 8, 'n_estimators': 246, 'subsample': 0.8}
MSE amb hiperparàmetres ajustats: 8.83265073478316


Per tant, podem considerar considerar el Gradient Boosting (amb els hiperparàmetres ajustats) com el millor model per les nostres dades, si ens basem en el MSE (8.83 en el Gradient Boosting i 11.49 en el Random Forest).

# Exercici 4

Compara el seu rendiment emprant l’aproximació traint/test o emprant totes les dades (validació interna).

Com que ja hem utilitzat l'aproximació train/test, ara utilitzarem la validació interna, per tal de comparar el rendiment del model. Tot i això, hem de tenir en compte que utilitzar totes les dades no és recomanable perquè podem sobreajustar el model (overfitting), de manera que tindrà molt bon rendiment amb les dades d'entrenament, però no serà bo amb noves dades que no ha vist anteriorment. La validació interna també pot fer que el model no sigui prou robust, ja que no podrem avaluar com varia el seu rendiment amb diferents subconjunts de dades. <br>
<br>
Utilitzarem els hiperparàmetres òptims que hem obtingut en el train/test i així farem la comparació del rendiment utilitzant la validació interna. En primer lloc per al Random Forest:

In [15]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Millors hiperparàmetres per a Random Forest
best_params_rf = {
    'max_depth': 8,
    'max_features': 'sqrt',
    'min_samples_leaf': 1,
    'min_samples_split': 2,
    'n_estimators': 200
}

# Creem el model de Random Forest amb els millors hiperparàmetres
rf_model = RandomForestRegressor(
    n_estimators=best_params_rf['n_estimators'],
    max_depth=best_params_rf['max_depth'],
    max_features=best_params_rf['max_features'],
    min_samples_split=best_params_rf['min_samples_split'],
    min_samples_leaf=best_params_rf['min_samples_leaf'],
    random_state=42
)

# Entrenem el model amb totes les dades
rf_model.fit(X, y)

# Fem prediccions amb les mateixes dades per a validació interna
y_pred_rf = rf_model.predict(X)

# Calculem l'MSE i el R2
mse_rf = mean_squared_error(y, y_pred_rf)
r2_rf = r2_score(y, y_pred_rf)

print(f'MSE amb validació interna per a Random Forest: {mse_rf}')
print(f'R2 amb validació interna per a Random Forest: {r2_rf}')


MSE amb validació interna per a Random Forest: 3.0368088040204797
R2 amb validació interna per a Random Forest: 0.9640271882216158


I ara pel model Gradient Boosting:

In [16]:
from sklearn.ensemble import GradientBoostingRegressor

# Millors hiperparàmetres per al Gradient Boosting
best_params_gb = {
    'learning_rate': 0.05,
    'max_depth': 3,
    'min_samples_leaf': 2,
    'min_samples_split': 2,
    'n_estimators': 200,
    'subsample': 0.8
}

# Creem el model de Gradient Boosting amb els millors hiperparàmetres
gb_model = GradientBoostingRegressor(
    n_estimators=best_params_gb['n_estimators'],
    max_depth=best_params_gb['max_depth'],
    learning_rate=best_params_gb['learning_rate'],
    min_samples_split=best_params_gb['min_samples_split'],
    min_samples_leaf=best_params_gb['min_samples_leaf'],
    subsample=best_params_gb['subsample'],
    random_state=42
)

# Entrenem el model amb totes les dades
gb_model.fit(X, y)

# Fem prediccions amb les mateixes dades per a la validació interna
y_pred_gb = gb_model.predict(X)

# Calculem l'MSE i el R2
mse_gb = mean_squared_error(y, y_pred_gb)
r2_gb = r2_score(y, y_pred_gb)

print(f'MSE amb validació interna per a Gradient Boosting: {mse_gb}')
print(f'R2 amb validació interna per a Gradient Boosting: {r2_gb}')


MSE amb validació interna per a Gradient Boosting: 2.1161960209668984
R2 amb validació interna per a Gradient Boosting: 0.9749323957940242


Els resultats de la validació interna que hem obtingut per al Random Forest i el Gradient Boosting suggereixen un rendiment excepcional amb valors molt baixos de MSE i molt alts d'R2. Malgrat tot, aquesta aparent alta eficàcia podria ser enganyosa a causa del risc de overfitting, ja que els models podrien haver après les dades d'entrenament incloent anomalies o soroll que no generalitzarien bé amb dades noves.

# Exercici 5

No facis servir la variable del nombre d'habitacions (RM) a l’hora de fer prediccions.

En primer lloc eliminarem la variable RM (nombre d'habitacions) del nostre dataset i després tornarem a dividir les dades en train/test.

In [18]:
from sklearn.model_selection import train_test_split

# Separem les característiques independents (X) sense la columna 'RM' i la variable dependent (y)
X = df_housing.drop(['MEDV', 'RM'], axis=1)
y = df_housing['MEDV']

# Dividim les dades en conjunts d'entrenament i de test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


Ara tornem a aplicar el model Random Forest:

In [19]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error

# Definim els millors hiperparàmetres trobats per Random Forest
rf_best_params = {
    'n_estimators': 200,
    'max_features': 'sqrt',
    'max_depth': 8,
    'min_samples_split': 2,
    'min_samples_leaf': 1
}

# Creem el model de Random Forest amb els millors hiperparàmetres
rf_model = RandomForestRegressor(**rf_best_params, random_state=42)

# Entrenem el model de Random Forest
rf_model.fit(X_train, y_train)

# Fem prediccions amb el model de Random Forest
rf_predictions = rf_model.predict(X_test)

# Avaluem el model de Random Forest
rf_mse = mean_squared_error(y_test, rf_predictions)
print(f'Random Forest MSE: {rf_mse}')


Random Forest MSE: 16.963544782959584


I ara el model Gradient Boosting:

In [20]:
from sklearn.ensemble import GradientBoostingRegressor

# Definim els millors hiperparàmetres trobats per Gradient Boosting
gb_best_params = {
    'n_estimators': 200,
    'max_depth': 3,
    'min_samples_split': 2,
    'min_samples_leaf': 2,
    'learning_rate': 0.05,
    'subsample': 0.8
}

# Creem el model de Gradient Boosting
gb_model = GradientBoostingRegressor(**gb_best_params, random_state=42)

# Entrenem el model de Gradient Boosting
gb_model.fit(X_train, y_train)

# Fem prediccions amb el model de Gradient Boosting
gb_predictions = gb_model.predict(X_test)

# Avaluem el model de Gradient Boosting
gb_mse = mean_squared_error(y_test, gb_predictions)
print(f'Gradient Boosting MSE: {gb_mse}')


Gradient Boosting MSE: 14.499206622474698


Per comparar, els resultats que havíem obtingut anteriorment sense eliminar la variable RM (nombre d'habitacions) eren: 

- Random Forest MSE amb hiperparàmetres ajustats: 11.487434801540934
- Gradient Boosting MSE amb hiperparàmetres ajustats: 8.83265073478316

Després d'eliminar la variable RM, els resultats són:

- Random Forest MSE sense RM: 16.963544782959584
- Gradient Boosting MSE sense RM: 14.499206622474698

Podem veure que en ambdós casos, el rendiment del model ha empitjorat (MSE ha augmentat) quan s'ha eliminat la variable RM del conjunt de dades. Això indica que la variable RM és una característica important per a la predicció del valor mitjà de les cases (MEDV) en aquest conjunt de dades. La pèrdua d'aquesta informació fa que els models siguin menys precisos en les seves prediccions. Així doncs, podem considerar la variable RM com un bon indicador del valor de l'habitatge en el mercat immobiliari. No tenir en compte aquesta variable pot reduït significativament la capacitat del model en captar la variança en el valor dels habitatges.